### PR TITLE
fix sevalue_to_native function support JSBNativeDataHolder convert

### PR DESCRIPTION
**Pull request by xbpiao**
Re: #

### Changelog

* fix sevalue_to_native function support JSBNativeDataHolder convert

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

test case  :

```typescript
        let url = "test";   // test.png
        let callback = (err: Error | null, asset: Asset) => {

            if (err)
                console.warn(err);

                if (asset instanceof ImageAsset) {
                let tex = new Texture2D();
                tex.reset({
                    width: asset.width,
                    height: asset.height,
                });
                tex.uploadData(asset.data);   // fix this line , in native code asset.data is JSBNativeDataHolder
                let sp = new SpriteFrame();
                sp.texture = tex;
                // use this SpriteFrame
            }
        };

        resources.load(url, Asset, callback);
```

### CODE REVIEW

**Comment by xbpiao** [comment]
test code :
```typescript
        let url = "test";   // test.png
        let callback = (err: Error | null, asset: Asset) => {

            if (err)
                console.warn(err);

                if (asset instanceof ImageAsset) {
                let tex = new Texture2D();
                tex.reset({
                    width: asset.width,
                    height: asset.height,
                });
                tex.uploadData(asset.data);   // fix this line , in native code asset.data._data is JSBNativeDataHolder
                let sp = new SpriteFrame();
                sp.texture = tex;
                // use this SpriteFrame
            }
        };

        resources.load(url, Asset, callback);
```

read simple-texture.jsb.ts:

```typescript
simpleTextureProto.uploadData = function (source, level = 0, arrayIndex = 0) {
    let data;
    if (source instanceof jsbWindow.HTMLCanvasElement) {
        // @ts-ignore
        data = source.data;
    } else if (source instanceof jsbWindow.HTMLImageElement) {
        // @ts-ignore
        data = source._data;                              // <------- JSBNativeDataHolder
    } else if (ArrayBuffer.isView(source)) {
        data = source.buffer;
    }
    oldUpdateDataFunc.call(this, data, level, arrayIndex);    // <------- call jsb_conversions.h  sevalue_to_native function convert
};
```

File: native/cocos/bindings/manual/jsb_conversions.h [hunk 1]
```diff
@@ -733,8 +734,13 @@ sevalue_to_native(const se::Value &from, T **to, se::Object * /*ctx*/) { // NOLI
     } else if (data->isTypedArray()) {
         data->getTypedArrayData(&tmp, nullptr);
     } else {
-        CC_ABORT(); // bad type
-        return false;
+        auto *dataHolder = static_cast<cc::JSBNativeDataHolder *>(data->getPrivateData());
```
Comment (dumganhar) [inline native/cocos/bindings/manual/jsb_conversions.h]: Why assume that the private data is always `JSBNativeDataHolder` type?
It's better to check the JSB type by 
```cpp
void *privateData = data->getPrivateData();
if (privateData != nullptr && data->_getClass() == __jsb_cc_JSBNativeDataHolder_class) {
    auto *dataHolder = static_cast<cc::JSBNativeDataHolder *>(privateData);
    tmp = dataHolder->getData();
} else {
    CC_ABORT(); // bad type
    return false;
}
```

Comment (xbpiao) [reply native/cocos/bindings/manual/jsb_conversions.h]: you better！！！

Comment (xbpiao) [reply native/cocos/bindings/manual/jsb_conversions.h]: but  **jsb_cocos_auto.h** include **jsb_conversions.h**  , circular reference!

Comment (xbpiao) [reply native/cocos/bindings/manual/jsb_conversions.h]: jsb_conversions.h
```c++

extern bool conver_native_data(se::Object *data, uint8_t **native_data);

template <typename T>
typename std::enable_if_t<!std::is_pointer<T>::value && std::is_arithmetic<T>::value, bool>
sevalue_to_native(const se::Value &from, T **to, se::Object * /*ctx*/) { // NOLINT(readability-identifier-naming)
    se::Object *data = from.toObject();
    uint8_t *tmp = nullptr;
    if (conver_native_data(data, &tmp) == false) {
        CC_ABORT(); // bad type
        return false;
    }
    *to = reinterpret_cast<T *>(tmp);
    return true;
}
```

add jsb_conversions.cpp

```c++
#include "bindings/manual/jsb_conversions.h"
#include "cocos/bindings/auto/jsb_cocos_auto.h"
#include "core/data/JSBNativeDataHolder.h"

bool conver_native_data(se::Object *data, uint8_t **native_data)
{
    bool result = false;
    if (data->isArrayBuffer()) {
        data->getArrayBufferData(native_data, nullptr);
        result = true;
    } else if (data->isTypedArray()) {
        data->getTypedArrayData(native_data, nullptr);
        result = true;
    } else {
        void *privateData = data->getPrivateData();
        if (privateData != nullptr && data->_getClass() == __jsb_cc_JSBNativeDataHolder_class) {
            auto* dataHolder = static_cast<cc::JSBNativeDataHolder*>(privateData);
            if (dataHolder != nullptr) {
                *native_data = dataHolder->getData();
                result = true;
            }
        }
    }
    return result;
}
```

It works, are you sure modify it this?

Comment (dumganhar) [reply native/cocos/bindings/manual/jsb_conversions.h]: T may be other primitive types like `uint32_t*`, `uint16_t*`, etc. So `conver_native_data` wrapper is not good enough.
I think we could use `extern se::Class * __jsb_cc_JSBNativeDataHolder_class ;` in jsb_conversion.h to resolve the circular reference.

Comment (xbpiao) [reply native/cocos/bindings/manual/jsb_conversions.h]: This commit is probably the smallest change, I'm not sure ```jsb_cocos_auto.h``` automatically generated！
move ```extern se::Class * __jsb_cc_JSBNativeDataHolder_class ;``` in ```jsb_conversion.h``` affect

Comment (dumganhar) [reply native/cocos/bindings/manual/jsb_conversions.h]: I don't mean that `move` the declaration from jsb_cocos_auto.h to `jsb_conversion.h`. 
I mean just add `extern se::Class * __jsb_cc_JSBNativeDataHolder_class;` in `jsb_conversion.h`. 
Could you have a try?

Comment (xbpiao) [reply native/cocos/bindings/manual/jsb_conversions.h]: thank you!

duplicate extern __jsb_cc_JSBNativeDataHolder_class to resolve the circular reference ```jsb_cocos_auto.h```

test It works!

**Review by dumganhar — APPROVED** [review]

**Comment by dumganhar** [comment]
@cocos-robot run test cases

File: native/cocos/bindings/manual/jsb_conversions.h (hunk 1 above)
Comment (dumganhar) [reply native/cocos/bindings/manual/jsb_conversions.h]: Good to know that it works. 👍

**Comment by dumganhar** [comment]
Merged. Thanks for the patch.

**Pull request merged**